internal/controllers/imagepolicy_controller.go

```diff
@@ -327,6 +331,15 @@ func (r *ImagePolicyReconciler) reconcile(ctx context.Context, sp *patch.SerialP
 	if oldObj.Status.LatestImage != obj.Status.LatestImage {
 		obj.Status.ObservedPreviousImage = oldObj.Status.LatestImage
 	}
+
+	// update the digest only when the tag changed or when the digest hasn't been set, yet (or has been reset manually)2
```

Comment:
> I'm not sure why we need to add another layer of abstraction that provides another form of immutability by not updating the digest even when the tag is mutated and points to a different digest.

Because there is no way for the reconciler to know if the underlying digest has changed other than fetching it. Otherwise we would hit the registry upon each reconciliation of ImageRepository and like 90% of the times or more it will just lead to the same digest being returned.

Tags should already usually not be mutated to begin with but since there's no technical way to prevent that, the digest (which is technically immutable) serves as a way to pin an image to a certain version.

The problem this PR (and the accompanying IAC one) is trying to solve is to have a way to reference the digest in image update automations. It's not the purpose of these PRs to provides users with a way to use mutable tags (like `latest`).

Reply:
> The problem this PR (and the accompanying IAC one) is trying to solve is to have a way to reference the digest in image update automations. It's not the purpose of these PRs to provides users with a way to use mutable tags (like latest).

Yes, this implements the mechanics to patch YAMLs using digests. For mutable tags we'll probably need to add a reconciler to ImagePolicy and `.spec.interval` that will refresh the digest, but this is out of scope for this PR.